### DISCUSSION TITLE
Examples of Applying a Function in Python that takes in multiple inputs and using it to create a new column within an Ibis Dataframe?

### DISCUSSION THREAD

**Discussion by YCat33** [ibis-project/ibis#8136] [Q&A] [answered]
I currently have a calculate_date_diff() function that takes in two dates and then calculates the difference. I am using the mutate functionality to apply this function but the output seems to be the same regardless of the input: 

date_calc = diff_types.mutate(string_date_diff=calculate_date_diff(diff_types.registration_date, diff_types.cc_open_date)) 

Would anyone be able to provide an example or guidance?

**Comment by gforsyth** [comment]
Hey @YCat33 -- can you post the contents of your diffing function?  This should be possible as you've written it if the function is using the ibis API to operate on the two columns.

**Reply by YCat33** [reply]
Here is the function:
```
def calculate_date_diff(date_expr_or_date1, date_expr_or_date2):
    """
    Calculate the absolute date difference in days between two date-like inputs.

    Parameters:
    - date_expr_or_date1: Input representing a date (StringColumn, string, or datetime object).
    - date_expr_or_date2: Input representing a date (StringColumn, string, or datetime object).

    Returns:
    - Absolute date difference in days.
    """
    # Parse date-like inputs to datetime objects
    try:
        if isinstance(date_expr_or_date1, ir.DateColumn):
            date1 = date_expr_or_date1.cast('timestamp').execute()[0]
        elif isinstance(date_expr_or_date1, ir.StringColumn):
            date1 = parser.parse(date_expr_or_date1.cast('string').execute()[0])
        else:
            raise ValueError(f"Invalid type for date_expr_or_date1: {type(date_expr_or_date1)}")

        if isinstance(date_expr_or_date2, ir.DateColumn):
            date2 = date_expr_or_date2.cast('timestamp').execute()[0]
        elif isinstance(date_expr_or_date2, ir.StringColumn):
            date2 = parser.parse(date_expr_or_date2.cast('string').execute()[0])
        else:
            raise ValueError(f"Invalid type for date_expr_or_date2: {type(date_expr_or_date2)}")
 # Calculate the absolute date difference
        date_diff = abs((date2 - date1).days)
        return date_diff
    except ValueError as e:
        raise ValueError(f"Error in calculate_date_diff: {e}")
```

**Reply by cpcloud** [reply]
Can you also show:

- Example input data
- Current output of calling this function on that data
- Expected output of calling this function on that data

It's hard to interpret the meaning of

> the output seems to be the same regardless of the input

without some data to look at, and the exact sequence of calls you're making to get the unexpected results.

**Reply by YCat33** [reply]
Here is an example of the input data:
<img width="532" alt="image" src="https://github.com/ibis-project/ibis/assets/115039992/5ea93762-ffbc-4ccb-8cbf-da0304845e7c">

How I am applying the aforementioned function:

`
date_calc = diff_types.mutate(string_date_diff=calculate_date_diff(diff_types.registration_date, diff_types.cc_open_date)) 
`

Output of Applying the function:

<img width="656" alt="image" src="https://github.com/ibis-project/ibis/assets/115039992/44471a0c-45d3-4a2c-8aef-07845b4fb73b">

**Reply by gforsyth** [reply]
Hey @YCat33 -- you're getting the same result because of your calls to `execute()[0]` -- that's executing the op and pulling the first value of the new column.  

I would recommend something like this:
cast both columns to `date` (if it's already a date then the backend will no-op this), then subtract the two newly casted columns, then, depending on what you want the output type to be, cast to that.

```python
[ins] In [13]: from ibis import _

[ins] In [14]: t
Out[14]: 
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ cc_open_date ┃ date_reg   ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ string       │ date       │
├──────────────┼────────────┤
│ 2012-09-12   │ 2013-09-13 │
│ 2003-08-31   │ 2005-12-31 │
└──────────────┴────────────┘

[ins] In [15]: t.mutate(
          ...:     string_date_diff=(_.date_reg.cast("date") - _.cc_open_date.cast("date")).cast("str")
          ...: )
Out[15]: 
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
┃ cc_open_date ┃ date_reg   ┃ string_date_diff ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━┩
│ string       │ date       │ string           │
├──────────────┼────────────┼──────────────────┤
│ 2012-09-12   │ 2013-09-13 │ 366              │
│ 2003-08-31   │ 2005-12-31 │ 853              │
└──────────────┴────────────┴──────────────────┘
```

and if you want to encapsulate that in a function, you could do something like:

```python

[ins] In [16]: def calc_date_diff(left, right):
          ...:     return left.cast("date") - right.cast("date")

[ins] In [17]: t.mutate(string_date_diff=calc_date_diff(_.date_reg, _.cc_open_date).cast("str"))
Out[17]: 
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
┃ cc_open_date ┃ date_reg   ┃ string_date_diff ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━┩
│ string       │ date       │ string           │
├──────────────┼────────────┼──────────────────┤
│ 2012-09-12   │ 2013-09-13 │ 366              │
│ 2003-08-31   │ 2005-12-31 │ 853              │
└──────────────┴────────────┴──────────────────┘
```

**Reply by cpcloud** [reply]
You can also use `x.delta(y, 'day')` to guarantee an integer result.

**Reply by YCat33** [reply]
What role does the "_" play?

I am encountering a "OperationNotDefinedError: No translation rule for <class 'ibis.expr.operations.temporal.DateDiff'>" issue when working with the following?


`
def calc_date_diff(start_date, end_date):
    return start_date.cast("date") - end_date.cast("date")
`

`
date_calc = diff_types.mutate(
                            string_date_diff=calc_date_diff(diff_types.registration_date, diff_types.cc_open_date).cast("int")
                            ) 
`

`
date_calc
`

**Reply by gforsyth** [reply]
Ahh, we should've asked.  Which backend are you using?

The underscore is a deferred operator, lots of info here: https://ibis-project.org/how-to/analytics/chain_expressions.html

**Reply by YCat33** [reply]
Trino.

**Reply by gforsyth** [reply]
Hmm, `DateDiff` should be available for Trino for Ibis versions >= 7.1.0 -- which version of Ibis do you have installed?

**Reply by YCat33** [reply]
Do you have an example of how DateDiff would work? I am running 7.2.0

**Reply by gforsyth** [reply]
It should work by just subtracting.  That said, interval types are a pain and there might be some issues there.  I would recommend you follow @cpcloud's advice and try using `.delta` instead.

**Reply by gforsyth** [reply, marked as answer]
This is using the Trino backend:

```python
[ins] In [13]: t
Out[13]: 
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ cc_open_date ┃ date_reg   ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ string       │ date       │
├──────────────┼────────────┤
│ 2012-09-12   │ 2013-09-13 │
│ 2003-08-31   │ 2005-12-31 │
└──────────────┴────────────┘

[ins] In [14]: t.mutate(date_diff=_.date_reg.cast("date").delta(_.cc_open_date.cast("date"), "day"))
Out[14]: 
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━┓
┃ cc_open_date ┃ date_reg   ┃ date_diff ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━┩
│ string       │ date       │ int64     │
├──────────────┼────────────┼───────────┤
│ 2012-09-12   │ 2013-09-13 │       366 │
│ 2003-08-31   │ 2005-12-31 │       853 │
└──────────────┴────────────┴───────────┘

[ins] In [15]: def calc_date_diff(left, right):
          ...:     return left.cast("date").delta(right.cast("date"), "day")

[ins] In [16]: t.mutate(date_diff=calc_date_diff(_.date_reg, _.cc_open_date))
Out[16]: 
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━┓
┃ cc_open_date ┃ date_reg   ┃ date_diff ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━┩
│ string       │ date       │ int64     │
├──────────────┼────────────┼───────────┤
│ 2012-09-12   │ 2013-09-13 │       366 │
│ 2003-08-31   │ 2005-12-31 │       853 │
└──────────────┴────────────┴───────────┘
```
